File: problema6/java/Digraph.java
// Problema 6
// GEORGE HARRISON ROCHA
// 11201810309

/*
Este programa implementa o algoritmo de Edmonds Karp e, consequentemente,
o algoritmo de busca e largura. O algoritmo de busca e largura foi
implementado com algumas modificações, entre elas: o método retorna true
se existe caminho entre s e t e false caso contrário. A fim de deixar mais
eficiente, o método bfs() retorna true assim que o vértice t é visitado,
não visitando, portanto, os vértices alcançáveis que ainda não foram visitados.

Para representar a rede residual, utilizou-se arcos virtuais como auxílio.

Primeiramente o programa lê os números 'n', 'm', 's' e 't', que denotam,
respectivamente, o número de vértices, o número de arcos, a fonte e o
sorvedouro. Logo em seguida, cria-se um digrafo ponderado e lê-se a
sequência de pares de vértices com o respectivo peso w. Finalmente,
usando a classe NetworkFlow, calcula-se o fluxo máximo e o fluxo
através de cada arco do digrafo. Enfim, imprime-se o fluxo máximo e os
arcos com seus respectivo fluxos.
*/

package tg.problema6.java; //! TIRAR ISSO ANTES DE ENVIAR
import java.util.Arrays;
import java.util.Comparator;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

/**
 * Classe Digrafo.
 */
public class Digraph {
    private int V;
    private int E;
    private Arc [] adj;

    /**
     * Construtor do Digrafo.
     * @param V o número de vértices
     */
    public Digraph (int V) {
        this.V = V;
        this.E = 0;
        adj = new Arc[V];
    }

    public int order() {
        return this.V;
    }

    public int size() {
        return this.E;
    }

    public Arc adj(int u) {
        return this.adj[u];
    }

    /**
     * Adiciona um arco ao digrafo, juntamente com seu "gêmeo" virtual.
     * @param e o arco
     */
    public void add_arc(Arc e) {
        e.next = adj[e.tail()];
        adj[e.tail()] = e;

        Arc va = new Arc(e.head(), e.tail(), e.flow, true);
        va.next = adj[va.tail()];
        adj[va.tail()] = va;

        e.twin = va;
        va.twin = e;

        this.E += 1;
    }

    /**
     * Cria um vetor contendo todos os arcos do digrafo.
     * @return um vetor de arcos
     */
    public Arc[] arcs () {
        Arc [] arcs = new Arc[E];

        int k = 0;
        for (int i = 0; i < V; i++) {
            for (Arc no = adj[i]; no != null; no = no.next) {
                if (!no.virtual()) {
                    arcs[k] = no;
                    k++;
                }
            }
        }
        Arrays.sort(arcs, new VerticesCompare());
        return arcs;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());  // Numero de vértices
        Digraph D = new Digraph(n);

        int m = Integer.parseInt(br.readLine());  // Número de arcos;

        int s = Integer.parseInt(br.readLine());  // Fonte
        int t = Integer.parseInt(br.readLine());  // Sorvedouro

        // Leitura dos arcos
        for (int i = 0; i < m; i++) {
            String e[] = br.readLine().split(" ");
            int x = Integer.parseInt(e[0]);
            int y = Integer.parseInt(e[1]);
            int w = Integer.parseInt(e[2]);
            D.add_arc(new Arc(x, y, w));
        }

        // Contrução da rede de fluxo
        NetworkFlow fm = new NetworkFlow(D, s, t);

        System.out.println(fm.maxumum_flow());

        Arc[] arcs = D.arcs();
        for (Arc e: arcs) {
            System.out.println(e);
        }
        br.close();
    }
}

/**
 * Classe arco
 */
class Arc {
    private int tail;
    private int head;
    private boolean virtual;

    int capacity;
    int flow;

    Arc next;
    Arc twin;

    // Construtor de um arco não virtual
    public Arc (int tail, int head, int capacity) {
        set_arc(tail, head, capacity);
        this.flow = 0;
        this.virtual = false;
    }

    // Construtor de um arco
    public Arc (int tail, int head, int capacity, boolean virtual) {
        set_arc(tail, head, capacity);
        this.virtual = virtual;
    }

    /**
     * Método auxiliar para "setar" propriedades comuns de arcos virtuais e arcos não virtuais.
     * @param tail a origem do arco
     * @param head o destino do arco
     * @param capacity a capacidade do arco
     */
    private void set_arc(int tail, int head, int capacity) {
        this.tail = tail;
        this.head = head;
        this.capacity = capacity;
        this.next = null;
        this.twin = null;
    }

    public int tail() {
        return this.tail;
    }

    public int head() {
        return this.head;
    }

    public boolean virtual() {
        return this.virtual;
    }

    @Override
    public String toString() {
        return tail + " " + head + " " + flow;
    }
}

/**
 * Classe que computa uma rede de fluxo.
 */
class NetworkFlow {
    private Digraph D;
    private int s;
    private int t;
    private boolean [] vis;
    private ArrayList<Arc> pred;
    private int maxumum_flow;

    /**
     * Construtor da rede de fluxo.
     * @param D o digrafo, onde os pesos dos arcos representam a capacidade
     * @param s a fonte
     * @param t o sorvedouro
     */
    public NetworkFlow (Digraph D, int s, int t) {
        this.D = D;
        this.s = s;
        this.t = t;
        vis = new boolean[D.order()];
        pred = new ArrayList<Arc>(D.order());
        this.maxumum_flow = edmonds_karp();
    }

    public int maxumum_flow() {
        return this.maxumum_flow;
    }

    /**
     * Método que extrai a capacidade mínima do caminho aumentador.
     * @param p o caminho aumentador
     * @return a capacidade mínima do caminho
     */
    private int min(ArrayList <Arc> p) {
        int menor = capacity(p.get(0));
        for (Arc e: p) {
            int atual = capacity(e);
            if (atual < menor) {
                menor = atual;
            }
        }
        return menor;
    }

    /**
     * Recupera o caminho encontrado pela bfs na rede residual.
     * @return um ArrayList contendo os arcos do caminho de s à t
     */
    private ArrayList <Arc> path() {
        ArrayList <Arc> path = new ArrayList <Arc>();

        Arc p = pred.get(t);
        while (p != pred.get(s)) {
            path.add(0, p);
            p = pred.get(p.tail());
        }

        return path;
    }

    /**
     * Implementação do algoritmo de Edmonds Karp.
     * @return um inteiro que representa o fluxo máximo do digrafo D
     */
    public int edmonds_karp() {
        int valor = 0;
        ArrayList <Arc> p;
        while (bfs()) {
            p = path();
            int delta = min(p);
            valor += delta;
            for (Arc e: p) {
                if (!e.virtual()) {
                    e.flow += delta;
                    e.twin.capacity += delta;
                } else {
                    e.twin.flow -= delta;
                    e.capacity -= delta;
                }
            }
        }
        return valor;
    }

    /**
     * Implementação do algoritmo de busca e largura,
     * adaptado para verificar se existe caminho de s a t na rede residual.
     * @return true se existe caminho entre s e t e false caso contrário
     */
    private boolean bfs () {
        pred.clear();
        for (int i = 0; i < D.order(); i++) {
            vis[i] = false;
            pred.add(null);
        }
        vis[s] = true;
        ArrayList <Integer> fila = new ArrayList <Integer>();
        fila.add(s);

        while (fila.size() > 0) {
            int u = fila.remove(0);
            for (Arc v = D.adj(u); v != null; v = v.next) {
                if (capacity(v) > 0) {
                    if (!vis[v.head()]) {
                        vis[v.head()] = true;
                        pred.set(v.head(), v);
                        fila.add(v.head());
                    }
                    if (vis[t]) {
                        return true;
                    }
                }
            }
        }
        return vis[t];
    }

    /**
     * Computa a capacidade do arco e na rede residual.
     * @param e o arco
     * @return um inteiro que representa a capacidade do arco e na rede residual
     */
    private int capacity(Arc e) {
        if (!e.virtual()) {
            return e.capacity - e.flow;
        } else {
            return e.twin.flow;
        }
    }
}

/**
 * Classe auxiliar usada para fazer a ordenação dos arcos.
 */
class VerticesCompare implements Comparator <Arc> {

    @Override
	public int compare(Arc e1, Arc e2) {
        if (e1.tail() < e2.tail()) {
			return -1;
		}
        if (e1.tail() > e2.tail()) {
			return 1;
		}

        if (e1.head() < e2.head()) {
        	return -1;
        }
        if (e1.head() > e2.head()) {
			return 1;
		}
        return 0;
    }
}

File: problema5/python/Digraph.py
# Problema 5
# George Harrison Rocha
# 11201810309

"""
Este programa implementa o algoritmo de dijkstra e a estrutura heap. 
A classe SSP implementa o algoritmo de dijkstra, e este, por sua vez, 
utiliza a classe Heap para manter a fila de prioridades. 

A classe Heap tem o atributo opcional 'key', que recebe uma função 
como valor. Esta função pega um elemento da heap e retorna um valor que 
então é usado para fazer as operações na heap. Caso não seja passado 
nenhuma função, o valor de retorno é o próprio elemento. No caso desse 
programa, a função passada é uma que pega um elemento x da heap retona 
o valor armazenado no índice x da lista 'cost' da classe SSP. 

Primeiramente, o programa lê os valores 'n', 'm', 'origin' e 'destination', 
que representam, respectivamente, a quantidade de vértices, a quantidade 
de conexões entre vértices, o vértice de origem e o vértice de destino. Logo em seguida, 
cria-se um digrafo ponderado e lê-se uma sequência de pares de vértices 
com o respectivo peso w. Dado o contexto do problema, para cada par u, v lido, 
adiciona-se ao digrafo um arco uv e um arco vu, ambos com peso w. Finalmente, 
usando a classe SSP, calcula-se o caminho mínimo entre 'origin' e 'destination', 
e imprime, caso exista caminho, a quantidade de vértices do caminho encontrado, 
o custo do caminho e os vértices do caminho. Caso não exista um caminho, 
imprime-se 'NO ROUTE FOUND!'
"""

from math import inf


class Arc:

    def __init__(self, head, tail, weight):
        self.__head = head
        self.__tail = tail
        self.__weight = weight

    def head(self):
        return self.__head

    def tail(self):
        return self.__tail

    def weight(self):
        return self.__weight


class WDigraph:
    def __init__(self, V):
        self.__V = V
        self.__E = 0
        self.__adj = [[] for _ in range(V)]  # Lista de adjacencia

    def order(self):
        return self.__V

    def size(self):
        return self.__E

    def add_arc(self, e):
        self.__adj[e.tail()].append(e)
        self.__E += 1

    def adj(self, u):
        return self.__adj[u]


class SSP:
    __UNREACHABLE = -1

    def __init__(self, D, root=0):
        self.__D = D
        self.__root = root
        self.__cost = [inf] * D.order()
        self.__pred = [self.__UNREACHABLE] * D.order()
        self.__dijkstra()

    def __relax(self, e, Q):
        u = e.tail()
        v = e.head()
        if self.__cost[v] > self.__cost[u] + e.weight():
            self.__cost[v] = self.__cost[u] + e.weight()
            self.__pred[v] = u
            Q.up_heapify(Q.A().index(v))

    def __dijkstra(self):
        Q = Heap(self.__D.order(), key=lambda x: self.__cost[x])
        Q.init()

        self.__cost[self.__root] = 0
        self.__pred[self.__root] = self.__root
        Q.up_heapify(Q.A().index(self.__root))

        while Q.heap_size():
            u = Q.extract_min()
            if self.__cost[u] == inf:
                return
            for e in self.__D.adj(u):
                self.__relax(e, Q)

    def dist_to(self, u):
        return self.__cost[u]

    def has_path_to(self, u):
        return self.__cost[u] < inf

    def path_to(self, u):
        if not self.has_path_to(u):
            return None

        path = [u]

        if self.__root == u:
            return path

        p = self.__pred[u]
        print(self.__pred)
        print(self.__cost)
        while p != self.__root:
            # print('p', p, 'root', self.__root, 'pred[p]', self.__pred[p])
            path.insert(0, p)
            p = self.__pred[p]
        path.insert(0, p)

        return path


class Heap:
    def __init__(self, n, key=None):
        self.__A = []
        self.__heap_size = 0
        self.__size = n
        self.__key = key if key is not None else lambda x: x

    def init(self):
        self.__A = list(range(self.__size))
        self.__heap_size = self.__size

    def heap_size(self):
        return self.__heap_size

    def A(self, i=-1):
        return self.__A[i] if i != -1 else self.__A

    @staticmethod
    def __left_child(i):
        return (2 * i) + 1

    @staticmethod
    def __right_child(i):
        return (2 * i) + 2

    @staticmethod
    def __parent(i):
        return (i - 1) // 2

    def __min_heapify(self, i):
        l = self.__left_child(i)
        r = self.__right_child(i)

        min = None
        if l < self.heap_size() and self.__key(self.A(l)) < self.__key(self.A(i)):
            min = l
        else:
            min = i
        if r < self.heap_size() and self.__key(self.A(r)) < self.__key(self.A(min)):
            min = r

        if min != i:
            self.__A[i], self.__A[min] = self.A(min), self.A(i)
            self.__min_heapify(min)

    def extract_min(self):
        min = self.A(0)
        self.__A[0] = self.A(self.heap_size() - 1)
        self.__heap_size -= 1
        self.__min_heapify(0)
        return min

    def up_heapify(self, i):
        while i > 0 and self.__key(self.A(self.__parent(i))) > self.__key(self.A(i)):
            self.__A[i], self.__A[self.__parent(i)] = self.A(self.__parent(i)), self.A(i)
            i = self.__parent(i)


input()
n = int(input()[15:])
m = int(input()[10:])
origin = int(input()[7:])
destination = int(input()[12:])
input()

D = WDigraph(n)

for _ in range(m):
    x, y, w = input().split(' ')
    x, y, w = int(x), int(y), float(w)
    D.add_arc(Arc(x, y, w))
    D.add_arc(Arc(y, x, w))

ssp = SSP(D, origin)
route = ssp.path_to(destination)

if route is None:
    print('NO ROUTE FOUND!')
else:
    total_jumps = len(route)
    print('TOTAL JUMPS:', total_jumps)
    print(f'ENERGY REQUIRED: {ssp.dist_to(destination):.4f}')
    print('ROUTE:')
    for i, v in enumerate(route):
        print(f'{i}: {v}')
